src/dataflow/operators/notificator.rs

```diff
@@ -0,0 +1,136 @@
+    pending: Vec<Capability<T>>,
+    frontier: Vec<MutableAntichain<T>>,
+    available: VecDeque<(Capability<T>, i64)>,
+}
+
+impl<T: Timestamp> Notificator<T> {
+    pub fn new() -> Notificator<T> {
+        Notificator {
+            pending: Vec::new(),
+            frontier: Vec::new(),
+            available: VecDeque::new(),
+        }
+    }
+
+    /// Updates the `Notificator`'s frontiers from a `CountMap` per input.
+    pub fn update_frontier_from_cm(&mut self, count_map: &mut [CountMap<T>]) {
+        while self.frontier.len() < count_map.len() {
+            self.frontier.push(MutableAntichain::new());
+        }
+
+        for (index, counts) in count_map.iter_mut().enumerate() {
+            while let Some((time, delta)) = counts.pop() {
+                self.frontier[index].update(&time, delta);
+            }
+        }
+    }
+
+    /// Reveals the elements in the frontier of the indicated input.
+    pub fn frontier(&self, input: usize) -> &[T] {
+        self.frontier[input].elements()
+    }
+
+    /// Requests a notification at the time associated with capability `cap`. Takes ownership of
+    /// the capability.
+    ///
+    /// In order to request a notification at future timestamp, obtain a capability for the new
+    /// timestamp first, as show in the example.
+    ///
+    /// #Examples
+    /// ```
+    /// use timely::dataflow::operators::{ToStream, Unary};
+    /// use timely::dataflow::channels::pact::Pipeline;
+    ///
+    /// timely::example(|scope| {
+    ///     (0..10).to_stream(scope)
+    ///            .unary_notify(Pipeline, "example", Vec::new(), |input, output, notificator| {
+    ///                while let Some((cap, data)) = input.next() {
+    ///                    output.session(&cap).give_content(data);
+    ///                    let mut time = cap.time();
+    ///                    time.inner += 1;
+    ///                    notificator.notify_at(cap.delayed(&time));
+    ///                }
+    ///                while let Some((cap, count)) = notificator.next() {
+    ///                    println!("done with time: {:?}", cap.time());
+    ///                }
+    ///            });
+    /// });
+    /// ```
+    #[inline]
+    pub fn notify_at(&mut self, cap: Capability<T>) {
+        self.pending.push(cap);
+    }
+
+    /// Repeatedly calls `logic` till exhaustion of the available notifications.
+    ///
+    /// `logic` receives a capability for `t`, the timestamp being notified and a `count`
+    /// representing how many capabilities were requested for that specific timestamp.
+    #[inline]
+    pub fn for_each<F: FnMut(Capability<T>, i64)>(&mut self, mut logic: F) {
+        while let Some((cap, count)) = self.next() {
+            ::logging::log(&::logging::GUARDED_PROGRESS, true);
+            logic(cap, count);
+            ::logging::log(&::logging::GUARDED_PROGRESS, false);
+        }
+    }
+
+    #[inline]
+    fn scan_for_available_notifications(&mut self) {
```

Comment:
> If you want to be lazy, the notificator I wrote when trying this just had pending and frontier; no available. It just scanned the list on each call to next, which I feel slightly---but not too---bad about.

Were you somehow checking to ensure you're delivering one of the notifications in the minimal antichain or was it simply the first one you encountered that was below the frontier?

> I think the advantage to doing everything in order is that operators capable of compacting state know that when they get a notification they can compact times less than that of the notification; neither messages nor notifications will be arriving. Doing notifications out-of-order means they would have to track a notification frontier too.

Yup, I also talked with Zaheer (as a "user") and while he doesn't think he has written anything that relies on this:
* it certainly seems like that operators that keep inter-epoch state (sliding window and similar) need to have more complex logic if no guarantees are given on notification delivery order;
* we both believe that it may be better (as in: less surprising to newcomers) if the ordering was guaranteed;
I'll ask around in the meeting as well to know what people think the current guarantees are.

An example of operators that rely on ordering (albeit possibly not the most interesting one) is https://github.com/frankmcsherry/timely-dataflow/blob/master/src/dataflow/operators/queue.rs#L24

Considering that the changes in user-level code are minimal for the switch to capabilities, maybe we should just make the ordering guarantee explicit for this first iteration of the capability-based api to avoid subtly breaking a bunch of code? And then maybe (maybe) provide different notificators (in the scenario with completely user-level notificators) depending on the guarantees that the user is looking for?
My first re-implementation preserved the original semantics by using the non-capability-aware logic and stashing capabilities on the side: https://github.com/utaal/timely-dataflow/blob/capabilities-with-ordered-notifications/src/dataflow/operators/notificator.rs (and, yes, `LazyKeyedSet` needs a new name).